File: backend/app/schemas/analysis.py
"""
Pydantic v2 validation and serialization schemas for video analysis persistence.
Phase 10: Database Integration.
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.counting import CountingLineSchema
from app.schemas.lane_analysis import LaneRegionSchema


class AnalysisRunRequest(BaseModel):

    """Request payload to initiate a persisted video analysis run."""
    model_config = ConfigDict(extra="ignore")

    analysis_type: str = Field(
        default="full_pipeline",
        description="Type of analysis run: 'full_pipeline', 'analytics', 'lane_analysis', or 'counting'",
    )
    confidence_threshold: Optional[float] = Field(
        default=None,
        ge=0.01,
        le=1.0,
        description="YOLO detection confidence threshold (defaults to settings/model default)",
    )
    processing_fps: Optional[int] = Field(
        default=None,
        ge=1,
        le=30,
        description="Frame extraction sampling rate (defaults to settings.processing_fps)",
    )
    max_frames: Optional[int] = Field(
        default=None,
        ge=1,
        le=300,
        description="Maximum frames to process in this run",
    )
    iou_threshold: Optional[float] = Field(
        default=None,
        ge=0.05,
        le=0.95,
        description="ByteTrack Kalman IoU association threshold",
    )
    counting_line: Optional[CountingLineSchema] = Field(
        default=None,
        description="Virtual tripwire geometry configuration for line-crossing counting",
    )
    lanes: Optional[List[LaneRegionSchema]] = Field(
        default=None,
        description="Polygonal lane regions for lane assignment and density estimation",
    )
    persistence_threshold: Optional[int] = Field(
        default=2,
        ge=1,
        le=10,
        description="Temporal persistence threshold (consecutive frames required for lane assignment)",
    )


class CrossingEventSchema(BaseModel):
    """Schema for an individual persisted vehicle line-crossing event."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    track_id: int
    class_name: str
    direction: str
    frame_index: int
    timestamp_seconds: float
    centroid_x: float
    centroid_y: float
    line_label: str
    created_at: datetime


class TrafficMetricsRecordSchema(BaseModel):
    """Schema for persisted traffic flow metrics and time-series distributions."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    observation_duration_seconds: float
    total_volume: int
    flow_rate_per_minute: float
    flow_rate_per_hour: float
    is_extrapolated: bool
    class_distribution: Optional[List[Dict[str, Any]]] = None
    direction_distribution: Optional[List[Dict[str, Any]]] = None
    time_series_buckets: Optional[List[Dict[str, Any]]] = None
    created_at: datetime


class LaneResultRecordSchema(BaseModel):
    """Schema for persisted per-lane occupancy, vehicles, and image-space density."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    lane_id: str
    lane_name: str
    direction_hint: Optional[str] = None
    polygon_json: Optional[List[List[float]]] = None
    polygon_area_px2: float
    unique_vehicles_count: int
    peak_occupancy: int
    average_occupancy: float
    image_space_density: float
    normalized_density_score: float
    vehicle_class_counts: Optional[Dict[str, int]] = None
    density_unit: str = "vehicles/px²"
    density_calibration_warning: Optional[str] = None
    created_at: datetime


class AnalysisSessionSummarySchema(BaseModel):
    """Summary schema for an analysis execution run."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    video_id: str
    video_filename: Optional[str] = None
    analysis_type: str
    status: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    processing_time_ms: Optional[float] = None
    total_frames_processed: int
    total_vehicles_detected: int
    total_vehicles_counted: int
    error_message: Optional[str] = None


class AnalysisSessionDetailResponse(BaseModel):
    """Comprehensive detail response for an analysis session including child results."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    video_id: str
    video_filename: Optional[str] = None
    analysis_type: str
    status: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    processing_time_ms: Optional[float] = None
    total_frames_processed: int
    total_vehicles_detected: int
    total_vehicles_counted: int
    config_snapshot: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    traffic_metrics: Optional[TrafficMetricsRecordSchema] = None
    lane_results: List[LaneResultRecordSchema] = Field(default_factory=list)
    crossing_events: List[CrossingEventSchema] = Field(default_factory=list)


class AnalysisSessionListResponse(BaseModel):
    """Paginated list of analysis execution sessions."""
    model_config = ConfigDict(from_attributes=True)

    total: int
    limit: int
    offset: int
    sessions: List[AnalysisSessionSummarySchema]


class AnalysisInfoResponse(BaseModel):
    """Database persistence and schema metadata."""
    service_name: str = "AnalysisPersistenceEngine"
    version: str = "0.1.0"
    persisted_entities: List[str] = Field(
        default_factory=lambda: [
            "analysis_sessions",
            "traffic_metrics",
            "lane_results",
            "crossing_events",
        ]
    )
    supported_analysis_types: List[str] = Field(
        default_factory=lambda: [
            "full_pipeline",
            "analytics",
            "lane_analysis",
            "counting",
            "tracking",
            "detection",
        ]
    )
    deduplication_policy: str = "Enforced unique constraint (analysis_session_id, track_id, line_label) in database."
    extrapolation_policy: str = "is_extrapolated boolean flag persisted directly in traffic_metrics table."
    density_policy: str = "Image-space density (veh/px²) persisted with explicit calibration warning."

File: frontend/src/types/analysis.ts
/**
 * TypeScript interface definitions for Analysis Persistence and History.
 * Matches backend schemas in backend/app/schemas/analysis.py.
 */

import { TimeSeriesBucket } from './analytics';

export interface ClassMetric {
  class_name: string;
  count: number;
  percentage: number;
}

export interface DirectionMetric {
  direction: string;
  count: number;
  percentage: number;
}

export interface TrafficMetricsRecord {
  id: string;
  analysis_session_id: string;
  observation_duration_seconds: number;
  total_volume: number;
  flow_rate_per_minute: number;
  flow_rate_per_hour: number;
  is_extrapolated: boolean;
  class_distribution: ClassMetric[];
  direction_distribution: DirectionMetric[];
  time_series_buckets: TimeSeriesBucket[];
  created_at: string;
}

export interface LaneResultRecord {
  id: string;
  analysis_session_id: string;
  lane_id: string;
  lane_name: string;
  direction_hint?: string | null;
  polygon_json: number[][];
  polygon_area_px2: number;
  unique_vehicles_count: number;
  peak_occupancy: number;
  average_occupancy: number;
  image_space_density: number;
  normalized_density_score: number;
  vehicle_class_counts?: Record<string, number> | null;
  density_unit: string;
  density_calibration_warning?: string | null;
  created_at: string;
}

export interface CrossingEventRecord {
  id: string;
  analysis_session_id: string;
  track_id: number;
  class_name: string;
  direction: string;
  frame_index: number;
  timestamp_seconds: number;
  centroid_x: number;
  centroid_y: number;
  line_label: string;
  created_at: string;
}

export interface AnalysisSessionSummary {
  id: string;
  video_id: string;
  video_filename?: string | null;
  analysis_type: string;
  status: string;
  started_at: string;
  completed_at?: string | null;
  processing_time_ms?: number | null;
  total_frames_processed: number;
  total_vehicles_detected: number;
  total_vehicles_counted: number;
  error_message?: string | null;
}

export interface AnalysisSessionDetail extends AnalysisSessionSummary {
  config_snapshot?: Record<string, any> | null;
  traffic_metrics?: TrafficMetricsRecord | null;
  lane_results: LaneResultRecord[];
  crossing_events: CrossingEventRecord[];
}

export interface AnalysisSessionListResponse {
  total: number;
  limit: number;
  offset: number;
  sessions: AnalysisSessionSummary[];
}

export interface AnalysisInfoResponse {
  service_name: string;
  version: string;
  persisted_entities: string[];
  supported_analysis_types: string[];
  deduplication_policy: string;
  extrapolation_policy: string;
  density_policy: string;
}

// =========================================================================
// Phase 17: Analysis Job Orchestration Types
// =========================================================================

export type AnalysisJobStatus = 'queued' | 'running' | 'completed' | 'failed' | 'cancelled';

export interface AnalysisJob {
  id: string;
  video_id: string;
  session_id?: string | null;
  status: AnalysisJobStatus;
  analysis_type: string;
  progress?: number | null;
  frames_processed: number;
  total_frames?: number | null;
  processing_fps?: number | null;
  cancellation_requested: boolean;
  error_code?: string | null;
  error_message?: string | null;
  provenance_category: string;
  is_synthetic: boolean;
  config_snapshot?: Record<string, any> | null;
  created_at: string;
  started_at?: string | null;
  completed_at?: string | null;
  updated_at: string;
}

export interface AnalysisJobCreateRequest {
  video_id: string;
  analysis_type?: string;
  confidence_threshold?: number;
  processing_fps?: number;
  max_frames?: number;
  iou_threshold?: number;
  counting_line?: any;
  lanes?: any[];
  persistence_threshold?: number;
}

export interface AnalysisJobListResponse {
  total: number;
  limit: number;
  offset: number;
  jobs: AnalysisJob[];
}

export interface AnalysisJobCancelResponse {
  status: string;
  message: string;
  job: AnalysisJob;
}
